kroxylicious-runtime/src/main/java/io/kroxylicious/proxy/internal/KafkaProxyExceptionHandler.java

```diff
@@ -0,0 +1,564 @@
+import org.apache.kafka.common.requests.EndQuorumEpochRequest;
+import org.apache.kafka.common.requests.EndTxnRequest;
+import org.apache.kafka.common.requests.EnvelopeRequest;
+import org.apache.kafka.common.requests.ExpireDelegationTokenRequest;
+import org.apache.kafka.common.requests.FetchRequest;
+import org.apache.kafka.common.requests.FetchSnapshotRequest;
+import org.apache.kafka.common.requests.FindCoordinatorRequest;
+import org.apache.kafka.common.requests.GetTelemetrySubscriptionsRequest;
+import org.apache.kafka.common.requests.HeartbeatRequest;
+import org.apache.kafka.common.requests.IncrementalAlterConfigsRequest;
+import org.apache.kafka.common.requests.InitProducerIdRequest;
+import org.apache.kafka.common.requests.JoinGroupRequest;
+import org.apache.kafka.common.requests.LeaderAndIsrRequest;
+import org.apache.kafka.common.requests.LeaveGroupRequest;
+import org.apache.kafka.common.requests.ListClientMetricsResourcesRequest;
+import org.apache.kafka.common.requests.ListGroupsRequest;
+import org.apache.kafka.common.requests.ListOffsetsRequest;
+import org.apache.kafka.common.requests.ListPartitionReassignmentsRequest;
+import org.apache.kafka.common.requests.ListTransactionsRequest;
+import org.apache.kafka.common.requests.MetadataRequest;
+import org.apache.kafka.common.requests.OffsetCommitRequest;
+import org.apache.kafka.common.requests.OffsetDeleteRequest;
+import org.apache.kafka.common.requests.OffsetFetchRequest;
+import org.apache.kafka.common.requests.OffsetsForLeaderEpochRequest;
+import org.apache.kafka.common.requests.ProduceRequest;
+import org.apache.kafka.common.requests.PushTelemetryRequest;
+import org.apache.kafka.common.requests.RenewDelegationTokenRequest;
+import org.apache.kafka.common.requests.SaslAuthenticateRequest;
+import org.apache.kafka.common.requests.SaslHandshakeRequest;
+import org.apache.kafka.common.requests.StopReplicaRequest;
+import org.apache.kafka.common.requests.SyncGroupRequest;
+import org.apache.kafka.common.requests.TxnOffsetCommitRequest;
+import org.apache.kafka.common.requests.UnregisterBrokerRequest;
+import org.apache.kafka.common.requests.UpdateFeaturesRequest;
+import org.apache.kafka.common.requests.VoteRequest;
+import org.apache.kafka.common.requests.WriteTxnMarkersRequest;
+import org.apache.kafka.common.resource.PatternType;
+import org.apache.kafka.common.resource.ResourcePatternFilter;
+import org.apache.kafka.common.resource.ResourceType;
+import org.apache.kafka.common.security.auth.KafkaPrincipal;
+
+import io.kroxylicious.proxy.frame.DecodedRequestFrame;
+
+public class KafkaProxyExceptionHandler {
+
+    private final Map<Class<? extends Exception>, Function<Throwable, Optional<?>>> responsesByExceptionType;
+
+    public KafkaProxyExceptionHandler() {
+        responsesByExceptionType = new ConcurrentHashMap<>();
+    }
+
+    public void registerExceptionResponse(Class<? extends Exception> exceptionClass, Function<Throwable, Optional<?>> responseFunction) {
```

Comment:
You can make things marginally nicer for callers by declaring it like this:

```suggestion
    public <T extends Throwable> void registerExceptionResponse(Class<T> exceptionClass, Function<T, Optional<?>> responseFunction) {
```

Even if you don't want to make it generic, the signature is weird (`Class<? extends Exception>`, but you require a `Function<Throwable, ...>`??)

Reply:
No, that seems like a sensible change.

Reply:
Maybe its my tired brain, but it breaks on the internal map that stores the mappings. 

I agree that it should be `extends Throwable` so have pushed that up. Will leave this open to revisit later.